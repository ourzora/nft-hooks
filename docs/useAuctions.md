This hook fetches data found on the blockchain for the given curator in the zora auction house.

This hook takes two arguments:
1. curators: list of curators to query for
2. approved: only return auctions that match the given state: true = approved, false = not approved, undefined = any status

```ts
/**
 * Fetches on-chain NFT auction data for the given curator
 *
 * @param curators
 * @param approved
 * @returns useNFTType hook results include loading, error, and chainNFT data.
 */

type useAuctions = {
  curators: string[];
  approved: true | false | undefined;
}

type useAuctionsResponse = {
  loading: boolean;
  error: ResponseError | undefined;
  data: ReserveAuctionPartialFragment[];
}

type ReserveAuctionPartialFragment = {
  /** The auction ID */
  id: Scalars['ID'];
  /** The originating token contract for this auction */
  tokenContract: Scalars['String'];
  /** The token ID for this auction */
  tokenId: Scalars['BigInt'];
  /** Whether or not the auction has been approved by the curator */
  approved: Scalars['Boolean'];
  /** The length of time the auction is intended to run for, after the first bid is made */
  duration: Scalars['BigInt'];
  /**
   * The expected end of auction timestamp, which can change if bids were placed
   * within the last 15 minutes, and is not set until the first bid is placed
   */
  expectedEndTimestamp?: Maybe<Scalars['BigInt']>;
  /** The time the first bid was placed */
  firstBidTime?: Maybe<Scalars['BigInt']>;
  /** The minimum price of the first bid */
  reservePrice: Scalars['BigInt'];
  /** The sale percentage to send to the curator */
  curatorFeePercentage: Scalars['Int'];
  /** The address that should receive the funds once the NFT is sold */
  tokenOwner: {id: address};
  /** The address of the auction's curator */
  curator: {id: address};
  /** The address of the ERC-20 currency to run the auction with, or 0x0 for ETH */
  auctionCurrency: {
    id: string;
    name: string;
    symbol: string;
    decimals: number;
  };
  status: ReserveAuctionStatus;
  /** The current bid on this auction */
  currentBid?: Maybe<ReserveAuctionBid>;
  /** The previous bids on this auction */
  previousBids?: Maybe<Array<InactiveReserveAuctionBid>>;
  /** The timestamp of the block the Transfer was created in */
  createdAtTimestamp: Scalars['BigInt'];
  /** The number of the block the Transfer was created in */
  createdAtBlockNumber: Scalars['BigInt'];
  /** The timestamp at which the auction end function was called */
  finalizedAtTimestamp?: Maybe<Scalars['BigInt']>;
  /** The block number at which the auction end function was called */
  finalizedAtBlockNumber?: Maybe<Scalars['BigInt']>;
```