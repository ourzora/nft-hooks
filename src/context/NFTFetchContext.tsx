import React, { useContext, useMemo } from 'react';
import { Networks, NetworkIDs } from '../constants/networks';
import { MediaFetchAgent } from '../fetcher/MediaFetchAgent';

export type FetchContext = InstanceType<typeof MediaFetchAgent>;

export const defaultFetchAgent = new MediaFetchAgent(Networks.MAINNET);

export const NFTFetchContext = React.createContext(defaultFetchAgent);

type NFTFetchConfigurationProps = {
  networkId: NetworkIDs;
  // TODO(iain): fix children type
  children: React.ReactNode;
};

export const NFTFetchConfiguration = ({
  networkId,
  children,
}: NFTFetchConfigurationProps) => {
  const lastFetchContext = useContext(NFTFetchContext);
  const fetchAgent = useMemo(() => {
    if (lastFetchContext.networkId === networkId) {
      return lastFetchContext;
    }
    return new MediaFetchAgent(networkId);
  }, [networkId]);
  return (
    <NFTFetchContext.Provider value={fetchAgent}>{children}</NFTFetchContext.Provider>
  );
};
