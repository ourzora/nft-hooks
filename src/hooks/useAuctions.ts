import { useContext } from 'react';
import { ReserveAuctionPartialWithMediaFragment } from '../graph-queries/zora-graph-types';
import useSWR, { SWRConfiguration } from 'swr';

import { NFTFetchContext } from '../context/NFTFetchContext';
import { onErrorRetry } from '../fetcher/ErrorUtils';

export type useAuctionHouseType = {
  loading: boolean;
  error?: string;
  data?: ReserveAuctionPartialWithMediaFragment[];
};

/**
 * Fetches on-chain NFT auction data for the given curator
 *
 * @param curators
 * @param approved
 * @returns useNFTType hook results include loading, error, and data (ReserveAuctionPartialFragment).
 */
export function useAuctions(
  curators: readonly string[] = [],
  approved: boolean | null = null,
  options: SWRConfiguration<ReserveAuctionPartialWithMediaFragment[]> = {}
): useAuctionHouseType {
  options.onErrorRetry = onErrorRetry;
  const fetcher = useContext(NFTFetchContext);
  const queryKey = JSON.stringify({ type: 'useAuctions', curators, approved });
  const { data, error } = useSWR<ReserveAuctionPartialWithMediaFragment[]>(
    queryKey,
    async (query: string) => {
      const { curators, approved } = JSON.parse(query);
      return await fetcher.fetchReserveAuctions(
        curators.map((curator: string) => curator.toLowerCase()),
        approved
      );
    },
    options
  );

  return {
    loading: !error && !data,
    error,
    data,
  };
}
