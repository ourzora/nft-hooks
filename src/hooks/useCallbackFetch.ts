import { useCallback, useContext, useEffect } from 'react';

import { NFTFetchContext } from '../context/NFTFetchContext';
import type { MediaFetchAgent } from '../fetcher/MediaFetchAgent';

/**
 * Combines callback and fetch react hooks to load data. Similar to SWR
 *
 * @param datapoint data argument to pass into the fetch function
 * @param fetchFn MediaFetchAgent generic fetch agent to pass into fetch function
 */
export function useCallbackFetch<T>(
  datapoint: T,
  fetchFn: (agent: MediaFetchAgent, data: NonNullable<T>) => void
) {
  const fetchAgent = useContext(NFTFetchContext);

  const fetch = useCallback(fetchFn, []);
  useEffect(() => {
    if (datapoint) {
      fetch(fetchAgent, datapoint as NonNullable<T>);
    }
  }, Array.isArray(datapoint) ? datapoint : [datapoint]);
}
