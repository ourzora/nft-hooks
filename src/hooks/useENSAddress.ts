import { useContext } from 'react';
import { ENSInfo } from '../backends/ens-reverse/ENSInterface';
import useSWR, { SWRConfiguration } from 'swr';

import { NFTFetchContext } from '../context/NFTFetchContext';

/**
 * useENSAddress - Load ens address for pretty display of user addresses
 *
 * @param address string address to fetch zora username of
 * @returns DomainResolvedPartFragment
 */
export function useENSAddress(address?: string, options?: SWRConfiguration<any>) {
  const { fetcher } = useContext(NFTFetchContext);
  return useSWR<ENSInfo>(
    address ? ['loadENS', address.toLowerCase()] : null,
    async (_, address: string) => fetcher.loadEnsName(address),
    options
  );
}
