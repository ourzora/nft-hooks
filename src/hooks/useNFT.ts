import { useContext } from 'react';
import useSWR from 'swr';

import { NFTFetchContext } from '../context/NFTFetchContext';
import { addAuctionInformation } from '../fetcher/TransformFetchResults';
import { getCurrenciesInUse } from '../fetcher/ExtractResultData';
import { NFTDataType } from '../fetcher/FetchResultTypes';

export type useNFTType = {
  currencyLoaded: boolean;
  error?: string;
  data?: NFTDataType;
};

type OptionsType = {
  refreshInterval?: number;
  initialData?: any;
  loadCurrencyInfo?: boolean;
};

/**
 * Fetches on-chain NFT data and pricing for the given zNFT id
 *
 * @param id id of zNFT to fetch blockchain information for
 * @param options SWR flags and an option to load currency info
 * @returns useNFTType hook results include loading, error, and chainNFT data.
 */
export function useNFT(id?: string, options: OptionsType = {}): useNFTType {
  const fetcher = useContext(NFTFetchContext);
  const { loadCurrencyInfo = false, refreshInterval, initialData } = options || {};

  const nftData = useSWR(
    id ? ['loadNFTDataUntransformed', id] : null,
    (_, id) => fetcher.loadNFTDataUntransformed(id),
    { refreshInterval, dedupingInterval: 0, initialData }
  );
  const currencyData = useSWR(
    nftData.data && loadCurrencyInfo
      ? ['loadCurrencies', ...getCurrenciesInUse(addAuctionInformation(nftData.data))]
      : null,
    (_, ...currencies) => fetcher.loadCurrencies(currencies),
    {
      refreshInterval,
      dedupingInterval: 0,
    }
  );

  let data;
  if (nftData.data !== undefined) {
    data = addAuctionInformation(nftData.data, currencyData.data);
  }

  return {
    currencyLoaded: !!currencyData.data,
    error: nftData.error,
    data,
  };
}
