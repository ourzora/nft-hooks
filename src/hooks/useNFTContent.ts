import { useContext } from 'react';
import useSWR, { SWRConfiguration } from 'swr';

import { MediaContentType } from '../fetcher/FetchResultTypes';
import { NFTFetchContext } from '../context/NFTFetchContext';

export type useNFTContentType = {
  error?: string;
  content?: MediaContentType;
};

/**
 * Hook to fetch NFT content from uri and mimetype
 * Fetches text mime types and returns uris for non-text media
 * that should be embedded
 *
 * @param uri URI of content to load or return URI for
 * @param mimeType MIME type expected for content
 * @returns useNFTContentType
 */
export function useNFTContent(
  uri?: string,
  mimeType?: string,
  options?: SWRConfiguration<MediaContentType>
): useNFTContentType {
  const fetcher = useContext(NFTFetchContext);
  const mimeTypeFetched = useSWR(
    uri && !mimeType ? ['fetchContentMimeType', uri] : null,
    (_, uri) => fetcher.fetchContentMimeType(uri)
  );
  const mimeTypeResult = mimeType || mimeTypeFetched.data;
  const content = useSWR(
    uri && mimeTypeResult ? ['fetchContent', uri, mimeTypeResult] : null,
    (_, uri, mimeTypeResult) => fetcher.fetchContent(uri, mimeTypeResult),
    options,
  );

  const error = mimeTypeFetched.error || content.error;
  return {
    error,
    content: content.data,
  };
}
