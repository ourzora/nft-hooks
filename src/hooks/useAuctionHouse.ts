import { useState } from 'react';
import { getCurrenciesInUse } from '../fetcher/ExtractResultData';

import { NFTDataType } from '../fetcher/FetchResultTypes';
import { useCallbackFetch } from './useCallbackFetch';

export type useAuctionHouseType = {
  loading: boolean;
  error?: string;
  data?: NFTDataType;
};

/**
 * Fetches on-chain NFT auction data for the given curator
 *
 * @param id id of zNFT to fetch blockchain information for
 * @param curator 
 * @returns useNFTType hook results include loading, error, and chainNFT data.
 */
export function useAuctionHouse(curator: string): useAuctionHouseType {
  const [data, setData] = useState<NFTDataType>();
  const [error, setError] = useState<string | undefined>();

  useCallbackFetch(curator, async (fetchAgent, curator) => {
    try {
      let data = await fetchAgent.loadAuctionByCurator(curator);
      setData(data);
    } catch (err) {
      setError(err.toString());
    }
  });

  return {
    loading: !error && !data,
    error,
    data,
  };
}
