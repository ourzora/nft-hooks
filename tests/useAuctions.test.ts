import { renderHook } from '@testing-library/react-hooks';

import { mockGraphQLQuery } from './setupZoraGQLMock';

import fetchMock, { getLastGraphQuery } from './setupFetchMock';

import { defaultFetchAgent } from '../src/context/NFTFetchContext';
import { useAuctions } from '../src/hooks/useAuctions';

let id = 0;
const getId = () => id++;

describe('useNFT', () => {
  beforeEach(() => {
    fetchMock.reset();
    defaultFetchAgent.clearCache();
  });
  const RESERVE_AUCTION_MOCK = {
    tokenId: getId,
    status: 'Active',
    approved: true,
    curatorFeePercentage: 10000,
  };

  it('loads all auctions', async () => {
    const mockOverrides = {
      ReserveAuction: () => ({
        ...RESERVE_AUCTION_MOCK,
      }),
    };

    mockGraphQLQuery(
      'https://api.thegraph.com/subgraphs/name/ourzora/zora-v1',
      mockOverrides
    );

    const { waitFor, result } = renderHook(() => useAuctions());

    await waitFor(() => result.current.loading === false);

    expect(
      getLastGraphQuery('https://api.thegraph.com/subgraphs/name/ourzora/zora-v1')
        .variables
    ).toEqual({
      first: 1000,
      skip: 0,
      approved: [true, false],
    });

    expect(result.current.error).toBeUndefined();
    expect(result.current.loading).toBeFalsy();
    expect(result.current.data).toMatchSnapshot();
  });

  it('loads specific auctions for a given curator', async () => {
    const mockOverrides = {
      ReserveAuction: () => ({
        ...RESERVE_AUCTION_MOCK,
      }),
    };

    mockGraphQLQuery(
      'https://api.thegraph.com/subgraphs/name/ourzora/zora-v1',
      mockOverrides
    );

    const { waitFor, result } = renderHook(() => useAuctions(['0xtestingcurator'], true));

    await waitFor(() => result.current.loading === false);

    expect(
      getLastGraphQuery('https://api.thegraph.com/subgraphs/name/ourzora/zora-v1')
        .variables
    ).toEqual({
      first: 1000,
      skip: 0,
      approved: [true],
      curators: ['0xtestingcurator'],
    });

    expect(result.current.error).toBeUndefined();
    expect(result.current.loading).toBeFalsy();
    expect(result.current.data).toMatchSnapshot();
  });
});
