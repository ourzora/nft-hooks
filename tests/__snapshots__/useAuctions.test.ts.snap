// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNFT loads all auctions 1`] = `
Array [
  Object {
    "approved": true,
    "approvedTimestamp": "12974",
    "auctionCurrency": Object {
      "decimals": 18,
      "id": "0xFACE",
      "name": "Wrapped Ether",
      "symbol": "WETH",
    },
    "createdAtTimestamp": "12974",
    "curator": Object {
      "id": "10",
    },
    "curatorFeePercentage": 10000,
    "currentBid": Object {
      "amount": "12974",
      "bidType": "Final",
      "bidder": Object {
        "id": "10",
      },
      "createdAtTimestamp": "12974",
      "transactionHash": "Hello World",
    },
    "duration": "12974",
    "expectedEndTimestamp": "12974",
    "finalizedAtTimestamp": "12974",
    "firstBidTime": "12974",
    "id": "0",
    "previousBids": Array [
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "3",
        "transactionHash": "Hello World",
      },
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "4",
        "transactionHash": "Hello World",
      },
    ],
    "reservePrice": "12974",
    "status": "Active",
    "token": "Hello World",
    "tokenContract": "Hello World",
    "tokenId": 0,
    "tokenOwner": Object {
      "id": "10",
    },
    "transactionHash": "Hello World",
  },
  Object {
    "approved": true,
    "approvedTimestamp": "12974",
    "auctionCurrency": Object {
      "decimals": 18,
      "id": "0xFACE",
      "name": "Wrapped Ether",
      "symbol": "WETH",
    },
    "createdAtTimestamp": "12974",
    "curator": Object {
      "id": "10",
    },
    "curatorFeePercentage": 10000,
    "currentBid": Object {
      "amount": "12974",
      "bidType": "Final",
      "bidder": Object {
        "id": "10",
      },
      "createdAtTimestamp": "12974",
      "transactionHash": "Hello World",
    },
    "duration": "12974",
    "expectedEndTimestamp": "12974",
    "finalizedAtTimestamp": "12974",
    "firstBidTime": "12974",
    "id": "1",
    "previousBids": Array [
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "6",
        "transactionHash": "Hello World",
      },
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "7",
        "transactionHash": "Hello World",
      },
    ],
    "reservePrice": "12974",
    "status": "Active",
    "token": "Hello World",
    "tokenContract": "Hello World",
    "tokenId": 1,
    "tokenOwner": Object {
      "id": "10",
    },
    "transactionHash": "Hello World",
  },
]
`;

exports[`useNFT loads specific auctions for a given curator 1`] = `
Array [
  Object {
    "approved": true,
    "approvedTimestamp": "12974",
    "auctionCurrency": Object {
      "decimals": 18,
      "id": "0xFACE",
      "name": "Wrapped Ether",
      "symbol": "WETH",
    },
    "createdAtTimestamp": "12974",
    "curator": Object {
      "id": "10",
    },
    "curatorFeePercentage": 10000,
    "currentBid": Object {
      "amount": "12974",
      "bidType": "Final",
      "bidder": Object {
        "id": "10",
      },
      "createdAtTimestamp": "12974",
      "transactionHash": "Hello World",
    },
    "duration": "12974",
    "expectedEndTimestamp": "12974",
    "finalizedAtTimestamp": "12974",
    "firstBidTime": "12974",
    "id": "0",
    "media": Object {
      "contentHash": "ByTeSStrInG",
      "contentURI": "Hello World",
      "createdAtTimestamp": "12974",
      "creator": Object {
        "id": "10",
      },
      "creatorBidShare": "12974",
      "currentAsk": Object {
        "amount": "12974",
        "createdAtTimestamp": "12974",
        "currency": Object {
          "decimals": 18,
          "id": "0xFACE",
          "name": "Wrapped Ether",
          "symbol": "WETH",
        },
        "id": "6",
      },
      "id": "5",
      "metadataHash": "ByTeSStrInG",
      "metadataURI": "Hello World",
      "owner": Object {
        "id": "10",
      },
      "ownerBidShare": "12974",
    },
    "previousBids": Array [
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "3",
        "transactionHash": "Hello World",
      },
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "4",
        "transactionHash": "Hello World",
      },
    ],
    "reservePrice": "12974",
    "status": "Active",
    "token": "Hello World",
    "tokenContract": "Hello World",
    "tokenId": 2,
    "tokenOwner": Object {
      "id": "10",
    },
    "transactionHash": "Hello World",
  },
  Object {
    "approved": true,
    "approvedTimestamp": "12974",
    "auctionCurrency": Object {
      "decimals": 18,
      "id": "0xFACE",
      "name": "Wrapped Ether",
      "symbol": "WETH",
    },
    "createdAtTimestamp": "12974",
    "curator": Object {
      "id": "10",
    },
    "curatorFeePercentage": 10000,
    "currentBid": Object {
      "amount": "12974",
      "bidType": "Final",
      "bidder": Object {
        "id": "10",
      },
      "createdAtTimestamp": "12974",
      "transactionHash": "Hello World",
    },
    "duration": "12974",
    "expectedEndTimestamp": "12974",
    "finalizedAtTimestamp": "12974",
    "firstBidTime": "12974",
    "id": "1",
    "media": Object {
      "contentHash": "ByTeSStrInG",
      "contentURI": "Hello World",
      "createdAtTimestamp": "12974",
      "creator": Object {
        "id": "10",
      },
      "creatorBidShare": "12974",
      "currentAsk": Object {
        "amount": "12974",
        "createdAtTimestamp": "12974",
        "currency": Object {
          "decimals": 18,
          "id": "0xFACE",
          "name": "Wrapped Ether",
          "symbol": "WETH",
        },
        "id": "11",
      },
      "id": "10",
      "metadataHash": "ByTeSStrInG",
      "metadataURI": "Hello World",
      "owner": Object {
        "id": "10",
      },
      "ownerBidShare": "12974",
    },
    "previousBids": Array [
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "8",
        "transactionHash": "Hello World",
      },
      Object {
        "amount": "12974",
        "bidInactivatedAtBlockNumber": "12974",
        "bidInactivatedAtTimestamp": "12974",
        "bidType": "Final",
        "bidder": Object {
          "id": "10",
        },
        "createdAtTimestamp": "12974",
        "id": "9",
        "transactionHash": "Hello World",
      },
    ],
    "reservePrice": "12974",
    "status": "Active",
    "token": "Hello World",
    "tokenContract": "Hello World",
    "tokenId": 3,
    "tokenOwner": Object {
      "id": "10",
    },
    "transactionHash": "Hello World",
  },
]
`;
